<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению. Сессия № 2
</center>
Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center> Домашнее задание № 1
## <center> Анализ данных по сердечно-сосудистым заболеваниям
*Идеальный просмотр Jupyter notebooks возможен только локально, GitHub и nbviewer могут неправильно отбражать формулы или картинки.*

**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным о сердечно-сосудистых заболеваниях (качать данные не надо, они уже есть в репозитории). Данные использовались в соревновании [Ml Boot Camp 5](http://mlbootcamp.ru/round/12/sandbox/).**

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/8D8uDf).**

В соревновании предлагалось определить наличие/отсутствие сердечно-сосудистых заболеваний (ССЗ) по результатам осмотра пациента.

**Описание данных.**

Объективные признаки:

 - Возраст (age)
 - Рост (height)
 - Вес (weight)
 - Пол (gender)
 

Результаты измерения:

 - Артериальное давление верхнее и нижнее (ap_hi, ap_lo)
 - Холестерин (cholesterol)
 - Глюкоза (gluc)
 

Субъективные признаки (со слов пациентов):

 - Курение (smoke)
 - Употребление алкоголя (alco)
 - Физическая активность (active)
 
Целевой признак (который интересно будет прогнозировать):
 - Наличие сердечно-сосудистых заболеваний по результатам классического врачебного осмотра (cardio)

Значения показателей холестерина и глюкозы представлены одним из трех классов: норма, выше нормы, значительно выше нормы. Значения субъективных признаков — бинарны.

Все показатели даны на момент осмотра.

Мы будем работать только с обучающей выборкой и с помощью `Pandas` проведем первичный анализ данных. 

Из библиотек нам понадобятся только `NumPy` и `Pandas`.

In [4]:
import numpy as np
import pandas as pd

Считываем данные из CSV-файла в объект pandas DataFrame.  

In [6]:
df = pd.read_csv('mlbootcamp5_train.csv', sep=';', 
                 index_col='id')


Посмотрим на первые 5 записей.

In [7]:
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168    62.0    110     80            1     1      0   
1   20228       1     156    85.0    140     90            3     1      0   
2   18857       1     165    64.0    130     70            3     1      0   
3   17623       2     169    82.0    150    100            1     1      0   
4   17474       1     156    56.0    100     60            1     1      0   

    alco  active  cardio  
id                        
0      0       1       0  
1      0       1       1  
2      0       0       1  
3      0       1       1  
4      0       0       0

**<font color='red'>Вопрос 1 (1 балл).</font> Сколько мужчин и женщин представлено в этом наборе данных? Не было дано расшифровки признака "пол" (какому полу соответствует 1, а какому – 2 в признаке ``gender``) – это определите, посмотрев также на рост при разумном предположении, что в среднем мужчины выше.**

**<font color='red'>Варианты:</font>**
- 45530 женщин и 24470 мужчин
- 45530 мужчин и 24470 женщин
- 45470 женщин и 24530 мужчин
- 45470 мужчин и 24530 женщин

In [16]:
df[df['gender'] == 1]['height'].mean(), \
df[df['gender'] == 2]['height'].mean()
# 1ый вариант


(161.35561168460356, 169.94789538210054)

**<font color='red'>Вопрос 2 (1 балл).</font> Кто в среднем чаще указывает, что употребляет алкоголь – мужчины или женщины?**

**<font color='red'>Варианты:</font>**
- мужчины 
- женщины

In [28]:
df[(df['gender'] == 2) & (df['alco'] == 1)]
# 1ый вариант

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                             
23     14532       2     181    95.0    130     90            1     1      1   
81     20330       2     187   115.0    130     90            1     1      0   
142    14499       2     164    48.0    110     70            1     2      1   
180    21473       2     174    90.0    140     80            1     1      1   
240    21848       2     170    69.0    120     80            1     1      1   
318    23376       2     175    75.0    180    100            3     1      1   
321    18444       2     173    75.0    130     80            2     1      1   
410    22415       2     173    70.0    130     70            2     1      1   
461    16069       2     160    85.0    120     80            1     1      0   
525    15215       2     172    71.0    160    120            3     3      1   
546    15912       2     171    74.0    120     80            1     1      1   
581    23262       2     162    70.0    130     70            1     1      1   
630    18978       2     175    86.0    120     80            1     1      1   
662    21993       2     175    92.0    120     80            3     1      1   
683    18825       2     164    68.0    140    100            1     1      0   
779    21159       2     160    55.0    120     80            1     1      1   
812    16488       2     175    95.0    120     90            1     1      1   
817    17420       2     175    88.0    140     90            2     1      1   
892    16018       2     166    58.0    125     80            1     1      1   
946    21094       2     167    90.0    130     85            2     2      1   
958    18265       2     168    75.0    110     80            1     1      1   
982    21101       2     170    90.0    110     70            1     1      1   
991    14549       2     165    85.0    120     80            1     1      1   
1032   21652       2     167    70.0    160     90            2     1      1   
1075   16208       2     179    87.0    140     78            1     1      1   
1194   23473       2     164    74.0    120     80            1     1      1   
1237   15297       2     170    65.0    120     80            1     1      1   
1275   22722       2     160    57.0    150     80            1     2      1   
1278   19890       2     163    71.0    130     80            1     1      1   
1310   19764       2     175    82.0    150    100            2     2      1   
...      ...     ...     ...     ...    ...    ...          ...   ...    ...   
98942  19073       2     175    67.0    140    100            1     1      1   
98945  15355       2     165   101.0    120     90            1     1      1   
98948  18709       2     169    76.0    130     80            1     2      1   
99039  17260       2     167    65.0    120     80            1     1      0   
99121  20230       2     165    72.0    140     90            1     1      1   
99251  16508       2     168    64.0    130     80            1     1      1   
99360  21023       2     177    85.0    110     80            1     1      0   
99401  18286       2     192    83.0    110     70            1     1      0   
99405  16781       2     179    93.0    180    100            1     2      0   
99412  21141       2     168    65.0    120     90            1     1      0   
99421  23334       2     182    89.0    140     90            3     1      0   
99437  23408       2     170    85.0    125     90            1     1      0   
99480  18377       2     156    60.0    110     90            1     1      1   
99551  23224       2     151    45.0    120     70            3     1      1   
99561  21116       2     170    68.0    115     70            1     1      0   
99640  16809       2     177    72.0    120     80            1     1      1   
99669  19250       2     158    72.0    140     90            1     1      0   
99686  19

**<font color='red'>Вопрос 3 (1 балл).</font> Во сколько раз (округленно, `round`) процент курящих среди мужчин больше, чем процент курящих среди женщин (по крайней мере, по этим анкетным данным)?**

**<font color='red'>Варианты:</font>**
- 4
- 8
- 12
- 16

In [6]:
# Ваш код здесь

**<font color='red'>Вопрос 4 (1 балл).</font> Вы наверняка заметили, что значения возраста какие-то странные. Догадайтесь, в чем здесь измеряется возраст, и ответьте, на сколько месяцев (примерно) отличаются медианные значения возраста курящих и некурящих.**

**<font color='red'>Варианты:</font>**
- 5
- 10
- 15
- 20

In [7]:
# Ваш код здесь

**<font color='red'>Вопрос 5 (2 балла).</font> В [статье](https://ru.wikipedia.org/wiki/Сердечно-сосудистый_риск) на Википедии про сердечно-сосудистый риск показана шкала SCORE для расчёта риска смерти от сердечно-сосудистого заболевания в ближайшие 10 лет. Вот она:**
<img src='../../img/SCORE2007.png' width=70%>

Давайте посмотрим на правый верхний прямоугольник, отображающий сегмент курящих мужчин в возрасте от 60 до 64 лет включительно. (Неочевидно, но тут для возраста и давления цифры означают верхнюю границу, и она не включается).

Видим 9-ку в левом нижнем углу этого прямоугольника и 47 – в правом верхнем. То есть если при этом систолическое (т.е. верхнее) артериальное давление – меньше 120 мм рт.ст., а уровень холестерина – 4 ммоль/л, то риск ССЗ оценивается примерно в 5 раз ниже, чем если бы давление лежало в интервале [160, 180), а холестерина было бы 8 ммоль/л.

Давайте посчитаем аналогичное значение, но на наших данных. 

Уточнения:
- Посчитайте признак ``age_years`` – возраст в годах, округлив до целых (`round`). Для данного примера отберите курящих мужчин от 60 до 64 лет включительно
- Категории уровня холестрина на рисунке и в наших данных отличаются. Отображение значений на картинке в значения признака ``cholesterol`` следующее: 4 ммоль/л $\rightarrow$ 1, 5-7 ммоль/л $\rightarrow$ 2, 8 ммоль/л $\rightarrow$ 3.
- Интересуют 2 подвыборки курящих мужчин возраста от 60 до 64 лет: первая с верхним артериальным давлением меньше 120 мм рт.ст. и концентрацией холестерина – 4 ммоль/л, а вторая – с верхним артериальным давлением от 160 до 180 мм рт.ст. (не включая верхнюю границу) и концентрацией холестерина – 8 ммоль/л.

**Во сколько раз (округленно, `round`) отличаются доли больных людей в этих двух подвыборках? Посчитайте на наших данных.**

**<font color='red'>Варианты:</font>**
- 2
- 3
- 4
- 5

In [8]:
# Ваш код здесь

**<font color='red'>Вопрос 6 (2 балла).</font> Постройте новый признак – BMI ([Body Mass Index](https://en.wikipedia.org/wiki/Body_mass_index)). Для этого надо вес в килограммах поделить на квадрат роста в метрах. Нормальными считаются значения BMI от 18.5 до 25. Выберите верные утверждения.**

**<font color='red'>Утверждения:</font>**
- Медианный BMI по выборке лежит в пределах нормы
- У женщин в среднем BMI выше, чем у мужчин
- У здоровых в среднем BMI выше, чем у больных
- В сегменте здоровых и непьющих мужчин BMI ближе к норме, чем в сегменте здоровых и непьющих женщин

In [9]:
# Ваш код здесь

**<font color='red'>Вопрос 7 (2 балла).</font> ** Можно заметить, что данные не особо-то чистые, много в них всякой "грязи" и неточностей. Еще лучше мы это увидим, когда обсудим визуализацию данных.

Отфильтруйте следующие сегменты пациентов (считаем это ошибками в данных)
- указанное нижнее значение артериального давления выше верхнего
- рост строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили (используйте `pd.Series.quantile`, если не знаете, что это такое – прочитайте)
- вес строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили

Этот вовсе не вся чистка данных, которую можно было проделать, но пока остановимся на этом.

**Сколько процентов данных (округленно, `round`) мы выбросили?**

**<font color='red'>Варианты:</font>**
- 8
- 9
- 10
- 11

In [10]:
# Ваш код здесь